### PR TITLE
[profiler][cupti] PM sampling engine

### PR DESCRIPTION
**Pull request by dolpm**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #187898
* #188939
* #188019
* __->__ #188849
* #188942

Add the standalone CUPTI PM-sampling engine (PmSampler): continuous sampling of the
on-chip GPU performance-monitor units -- e.g. SM-active % (sm__cycles_active),
DRAM-throughput % (gpu__dram_throughput), NVLink RX/TX % -- concurrently with the
activity monitor. PM sampling reads dedicated HW units, so it has negligible GPU-side
cost, but it locks the GPU clocks while active (which can shift absolute kernel
durations), so it is opt-in.

There is one PmSampler per CUDA device -- a per-device singleton: PmSampler(device)
(default the current device) returns the one instance, so PmSampler(0) is PmSampler(0),
and construction routes through __new__ (no separate factory). Only one PM session per
device is possible, and the process-global CUPTI profiler's deinit is not refcounted by
CUPTI (a second deinit segfaults), so PmSampler refcounts its live collectors and
deinitializes exactly once, when the last is torn down -- which makes sampling several
devices at once safe (profiler_initialize is idempotent). Users register as consumers:
add_consumer(metrics) returns a handle; handle.detach() drops it (dropping the handle
detaches too -- a GC safety net). The session samples the union of all consumers' metrics;
the first consumer starts it and the last one tears it down. Delivery is pull-based:
handle.poll() returns that consumer's
samples newer than its cursor, sliced to its own metrics. decode drains, so a sample
another consumer hasn't polled is retained (bounded to the look-back window) and GC'd once
every consumer has passed it -- a lone, caught-up consumer keeps no copy. A metric union
that cannot be collected in one PM pass is rejected at
add_consumer (PM sampling is single-pass), leaving any running session untouched; a union
change rebuilds the session (clean enable/configure/start) rather than reconfiguring in
place, since decode after an in-place stop/configure/start errors on some CUPTI versions.

The HW units sample autonomously into a device-side ring (KEEP_LATEST), so the engine
spawns no thread of its own -- collection is passive and each consumer drives handle.poll()
on its own cadence (the cupti_monitor backend polls from its flush thread). poll drains the
ring without disabling (safe to call repeatedly while sampling) and the final tail drain
happens when the last consumer leaves -- before disabling, since a wrapped ring is only
decodable while active. Each poll drains the ring (each sample once), so total decode cost
is independent of poll frequency; the only constraint is polling before the KEEP_LATEST ring
fills (span = max_samples * interval) and drops the oldest -- the caller picks the cadence
(the cupti_monitor backend uses its flush cadence). Interval + look-back are process-wide env
(TORCH_CUPTI_PM_SAMPLING_INTERVAL_MS default 1, TORCH_CUPTI_PM_SAMPLING_LOOKBACK_MS
default 10000) and size the ring exactly via CUPTI's pm_sampling_get_counter_data_size (no
per-sample byte estimate); metrics are NOT env -- each consumer brings its own.

Decoded samples are HW-timestamped in the CUPTI clock domain and delivered raw (start_ns /
device_id / one column per metric name) -- converting to trace/epoch time is the consumer's
job, so the engine has no clock dependency. The first sample of a session has a bogus
(uninitialized, nonzero) start timestamp, so samples outside the look-back of the newest
are dropped. The engine is thus usable on its own, independent of the profiler backend;
wiring it into the cupti_monitor backend lands in a follow-up.

Test Plan:
Real CUPTI PM sampling (needs a perfmon-capable GPU + cupti.pm_sampling; the tests skip
otherwise):

```
python test/profiler/test_cupti_pm_sampling.py
```

Validated on a GB200 (GB100 chip) -- all 14 tests pass with PM sampling actually engaged:
PmSampler(device) returns the per-device singleton, frames carry a column per metric, HW
timestamps are monotonic and within the look-back, max_samples matches look-back /
interval, metric selection works, two consumers share one session (sampling their union,
each sliced to its own metrics), a metric-union change rebuilds the session and sampling
survives it, the profiler-deinit refcount balances across an add/detach cycle, a dropped
handle still detaches, supported_metrics returns the chip's counters, an unknown metric
warns, an empty metric set and a multi-pass set (sm__throughput.*, ~8 passes) are rejected,
and the sampler retains samples only for a lagging consumer (a lone consumer keeps none).

Authored with the assistance of an AI coding assistant.

### CODE REVIEW

File: test/profiler/test_cupti_pm_sampling.py [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+
+
+# PM sampling goes through the CUPTI profiler API; gate it at the same >= 13.3 as the monitor.
+TEST_CUPTI_V13_3 = TEST_CUPTI_PYTHON and _cupti_version() >= 130300
+
+
+# Needs CUDA + libcupti >= 13.3 (cupti-python ships with PyTorch). Whether the HW can actually
+# engage (perfmon access) is only known at start(), so tests still skip at runtime when no session
+# comes up.
+TEST_CUPTI_PM_SAMPLING = TEST_CUDA and TEST_CUPTI_V13_3
```
Comment (ngimel) [inline test/profiler/test_cupti_pm_sampling.py]: do we reuse this? monitor tests should be gated on the same

**Review by ngimel — APPROVED** [review]

File: torch/profiler/_cupti/pm_sampling.py [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+    from collections.abc import Iterable
+    from typing_extensions import Self
+
+
+logger = logging.getLogger(__name__)
+
+# Interval + look-back are configured process-wide via env read at import (set them before importing
+# torch); metrics are NOT -- each consumer brings its own via add_consumer(). The HW sampling
+# interval (GPU_TIME_INTERVAL units = ns); TORCH_CUPTI_PM_SAMPLING_INTERVAL_MS, default 1 ms.
+_SAMPLING_INTERVAL_NS = int(
```
Comment (ngimel) [inline torch/profiler/_cupti/pm_sampling.py]: as per usual, can we set it via config and not env var?

Comment (dolpm) [reply torch/profiler/_cupti/pm_sampling.py]: the concern with setting it via. config is that it's global. i suppose we could have some sort of global config method, but it'd be first-come-first-serve.
